cvss/src/cvss4.rs

```diff
@@ -0,0 +1,510 @@
+    pub vi: VulnerableIntegrity,
+    pub va: VulnerableAvailability,
+    pub sc: SubsequentConfidentiality,
+    pub si: SubsequentIntegrity,
+    pub sa: SubsequentAvailability,
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum Cvss4Error {
+    Invalid,
+    MinorVersion,
+    AttackVector,
+    AttackComplexity,
+    AttackRequirements,
+    PrivilegesRequired,
+    UserInteraction,
+    VulnerableConfidentiality,
+    VulnerableIntegrity,
+    VulnerableAvailability,
+    SubsequentConfidentiality,
+    SubsequentIntegrity,
+    SubsequentAvailability,
+}
+
+impl FromStr for Cvss4Base {
+    type Err = Cvss4Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts = s.split('/').collect::<Vec<_>>();
+        if parts.len() == 12 && parts[0].starts_with("CVSS:4.0") {
+            let minor_version = 0;
+
+            let av = parts[1];
+            let ac = parts[2];
+            let at = parts[3];
+            let pr = parts[4];
+            let ui = parts[5];
+            let vc = parts[6];
+            let vi = parts[7];
+            let va = parts[8];
+            let sc = parts[9];
+            let si = parts[10];
+            let sa = parts[11];
+
+            Ok(Cvss4Base {
+                minor_version,
+                av: AttackVector::from_str(av)?,
+                ac: AttackComplexity::from_str(ac)?,
+                at: AttackRequirements::from_str(at)?,
+                pr: PrivilegesRequired::from_str(pr)?,
+                ui: UserInteraction::from_str(ui)?,
+                vc: VulnerableConfidentiality::from_str(vc)?,
+                vi: VulnerableIntegrity::from_str(vi)?,
+                va: VulnerableAvailability::from_str(va)?,
+                sc: SubsequentConfidentiality::from_str(sc)?,
+                si: SubsequentIntegrity::from_str(si)?,
+                sa: SubsequentAvailability::from_str(sa)?,
+            })
+        } else {
+            Err(Cvss4Error::Invalid)
+        }
+    }
+}
+
+impl Display for Cvss4Base {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "CVSS:4.{}/AV:{}/AC:{}/AT:{}/PR:{}/UI:{}/VC:{}/VI:{}/VA:{}/SC:{}/SI:{}/SA:{}",
+            self.minor_version,
+            self.av,
+            self.ac,
+            self.at,
+            self.pr,
+            self.ui,
+            self.vc,
+            self.vi,
+            self.va,
+            self.sc,
+            self.si,
+            self.sa,
+        )
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+pub enum AttackVector {
+    Network,
+    Adjacent,
+    Local,
+    Physical,
+}
+
+impl Display for AttackVector {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            match self {
+                Self::Network => 'N',
+                Self::Adjacent => 'A',
+                Self::Local => 'L',
+                Self::Physical => 'P',
+            }
+        )
+    }
+}
```

Comment:
Minor nit, but I like the local `use` to remove some of the match branch noise. (Ignore me if you were specifically using `Self::` for some other reason).

You have other places in this PR where you could do this, too. If you do it, only do it because you like it, not because I like it. :)

```suggestion
impl Display for AttackVector {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        use AttackVector::*;
        write!(
            f,
            "{}",
            match self {
                Network => 'N',
                Adjacent => 'A',
                Local => 'L',
                Physical => 'P',
            }
        )
    }
}
```

Reply:
That's a good idea, I just forget about use within fns.

lemme ammend.